Support Values in compiled runtime

henriknyman opened the pull request:
Based on #11395 Please wait for that to be merged and then rebase.


- Introduce tracking of ValueType and AnyValueType as representation types in CodeGenType
- Let parameters stay as type AnyValue
- Let property values stay as type Value
- Let literal lists and maps be AnyValue
- Handle AnyValue/Value types internally in a bunch of places
- Use the same conversion to storable type as the interpreted runtime, which handles
temporal and spatial types

NOTE: This is needed for the compiled runtime to properly handle the new temporal and spatial indexes.

Lojjs committed:
Add index tests for arrays of length > 1

henriknyman committed:
Support Values in compiled runtime

- Introduce tracking of ValueType and AnyValueType as
representation types in CodeGenType
- Let parameters stay as type AnyValue
- Let property values stay as type Value
- Let literal lists and maps be AnyValue
- Handle AnyValue/Value types internally in a bunch of places

henriknyman committed:
Unify behaviour of ArithmeticException across runtimes

- Math functions in the values module does not rethrow arithmetic
Java exceptions as Neo4j exceptions.
This lead to different behaviour across the runtimes, where
only the compiled runtime was doing this.
For now let us not alter the message (e.g. "long overflow") coming out
of Java, but just make sure we wrap it in a public Neo4j exception.

- Floating point division should not throw "/ by zero".
Division by zero has well-defined representations in the IEEE-754
floating point standard, and it would be rude to abort a query
unless the user has opted in to trapping floating point exceptional
conditions. Currently the JVM does not support such trapping.

henriknyman committed:
Fix some test failures

- Correct comparators for sorting AnyValue-fields
- Materialize node/relationship/path proxies when returning
list or map values
- Correct type for literal lists with mixed primitive types
represented as longs

henriknyman committed:
Fix regression in compiled toAnyValue

henriknyman committed:
Cleanup CompiledConversionUtils

- Fix ternary OR, NOT to support values
- Remove unused code
- Add unit tests

henriknyman committed:
Fix regression in GroupingAndAggreagtion benchmark

henriknyman committed:
Extend math helper test with Values

henriknyman committed:
Cleanup

henriknyman commented on community/codegen/pom.xml:
```diff
@@ -85,6 +85,11 @@ the relevant Commercial Agreement.
+      <dependency>
+          <groupId>org.neo4j</groupId>
+          <artifactId>neo4j-values</artifactId>
+          <version>3.4.0-alpha01</version>
```
Oops, should be `${project.version}`

henriknyman commented on enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/TemporalAcceptanceTest.scala:
```diff
@@ -134,6 +132,63 @@ class TemporalAcceptanceTest extends ExecutionEngineFunSuite with QueryStatistic
+//        plan should useOperatorWithText("NodeIndexSeek", ":Occasion(timeSpan)")
+//      }, expectPlansToFail = Configs.AbsolutelyAll - Configs.Version3_4 - Configs.Version3_3),
+//      params = Map("param" ->
+//        Array(LocalDate.of(2018, 4, 1), LocalDate.of(2018, 4, 2))))
```
Cleanup

henriknyman commented on enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/compiled/codegen/LogicalPlanConverter.scala:
```diff
@@ -197,16 +198,16 @@ object LogicalPlanConverter {
+//            val expressionVar = Variable(context.namer.newVarName(), CodeGenType.Any, nullable = false)
+//            ForEachExpression(expressionVar, expression,
+//                              indexSeekFun(opName, context.namer.newVarName(), LoadVariable(expressionVar), nodeVar,
+//                                           actions))
```
Cleanup

henriknyman commented on enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/compiled/codegen/ir/aggregation/Distinct.scala:
```diff
@@ -39,10 +39,12 @@ case class Distinct(opName: String, setName: String, vars: Iterable[(String, Cod
+//        if (expr.codeGenType == CodeGenType.Any) generator.assign(variable, expr.codeGenType,
+//                                                                  generator
+//                                                                    .materializeAny(expr.generateExpression(generator)))
+//        else generator.assign(variable, expr.codeGenType, expr.generateExpression(generator))
```
Cleanup

henriknyman commented on enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/compiled/helpers/LiteralTypeSupport.scala:
```diff
@@ -45,12 +45,48 @@ object LiteralTypeSupport {
+//  def deriveCodeGenTypeWithRepresentation(ct: CypherType, representationAs: RepresentationType): CypherCodeGenType = ct match {
+//    case ListType(innerCt) => CypherCodeGenType(CTList(innerCt), ListReferenceType(representationAs))
+//    case _ => CypherCodeGenType(ct, representationAs)
+//  }
```
Cleanup

henriknyman commented on enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_4/codegen/GeneratedMethodStructure.scala:
```diff
@@ -1323,9 +1391,10 @@ class GeneratedMethodStructure(val fields: Fields, val generator: CodeBlock, aux
 
   override def declareProperty(propertyVar: String) = {
-    val localVariable = generator.declare(typeRef[Object], propertyVar)
+//    val localVariable = generator.declare(typeRef[Object], propertyVar)
```
Cleanup

henriknyman committed:
Add temporal tests

henriknyman committed:
Fix issue with DISTINCT and nulls

Make sure any combination of null and NO_VALUE is considered the same

pontusmelke commented on community/codegen/src/main/java/org/neo4j/codegen/bytecode/ByteCodeExpressionVisitor.java:
```diff
@@ -325,7 +325,7 @@ public void notEqual( Expression lhs, Expression rhs )
     private void equal( Expression lhs, Expression rhs, boolean equal )
     {
-        assertSameType( lhs, rhs, "compare" );
+//        assertSameType( lhs, rhs, "compare" ); // TODO: Is this assert unnecessary / too strong?
```
In byte code you can only compare exact types so this assertion is valid I think

henriknyman replied:
I think this is true for some primitives since they have different compare instruction, but hardly for reference types? If I enable this assertion I cannot compare an `AnyValue` to a `Value`, but without it that yields correctly working bytecode.

pontusmelke commented on community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledConversionUtils.java:
```diff
@@ -44,76 +43,72 @@
-            for ( int i = 0; i < len; i++ )
+            SequenceValue sequenceValue = (SequenceValue) value;
+            Iterator<AnyValue> iterator = sequenceValue.iterator();
+            Set<AnyValue> set = sequenceValue.iterationPreference() == RANDOM_ACCESS ?
```
If `RANDOM_ACCESS` we can save on creating the iterator as well, so something like

```
if ( sequenceValue.iterationPreference() == RANDOM_ACCESS )
{
  for ( int i = 0; i < sequenceValue.length(); i++ )
  {
    set.add( sequenceValue.value( i ) );
  }
}
else
{
  //same as above
}
```

pontusmelke commented on community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledConversionUtils.java:
```diff
@@ -395,6 +444,10 @@ public static LongStream toLongStream( Object list )
         {
             return LongStream.empty();
         }
+        else if ( list instanceof SequenceValue )
```
can we get here?

henriknyman replied:
I don't think so (at least not in the current state of the code, it could have been the case at an earlier point during the development of this PR). We only get to this method from parameter loading, and now that parameters always comes as AnyValue and we do not convert them I don't think this method is even being called anymore.

pontusmelke commented on community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledConversionUtils.java:
```diff
@@ -395,6 +444,10 @@ public static LongStream toLongStream( Object list )
         }
+        else if ( list instanceof SequenceValue )
+        {
+            throw new IllegalArgumentException( "Need to implement support for SequenceValue in CompiledConversionUtils.toLongStream" );
```
can we get here?

henriknyman replied:
See above

pontusmelke commented on community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledConversionUtils.java:
```diff
@@ -431,6 +484,10 @@ public static DoubleStream toDoubleStream( Object list )
         }
+        else if ( list instanceof SequenceValue )
+        {
+            throw new IllegalArgumentException( "Need to implement support for SequenceValue in CompiledConversionUtils.toDoubleStream" );
```
can we get here?

henriknyman replied:
See above

pontusmelke commented on community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledConversionUtils.java:
```diff
@@ -458,6 +515,10 @@ public static IntStream toBooleanStream( Object list )
         }
+        else if ( list instanceof SequenceValue )
+        {
+            throw new IllegalArgumentException( "Need to implement support for SequenceValue in CompiledConversionUtils.toBooleanStream" );
```
can we get here?

pontusmelke commented on community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledMaterializeValueMapper.java:
```diff
@@ -0,0 +1,185 @@
+import org.neo4j.values.virtual.VirtualRelationshipValue;
+import org.neo4j.values.virtual.VirtualValues;
+
+public final class CompiledMaterializeValueMapper implements ValueMapper<AnyValue>
```
This looks overly complicated or am I missing something. It maps `AnyValue` to `AnyValue` via an internal ValueMapper that maps `AnyValue` to `Object`? Can't we just rely on EmbeddedProxySPI directly instead of going via the DefaultValueMapper?

pontusmelke commented on community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledMaterializeValueMapper.java:
```diff
@@ -0,0 +1,185 @@
+        {
+            return value;
+        }
+        return ValueUtils.fromNodeProxy( defaultValueMapper.mapNode( value ) );
```
couldn't we use the proxySPI here instead and do 
```
 return ValueUtils.fromNodeProxy( proxySpi.newNodeProxy( value.id() );
```

pontusmelke commented on community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledMaterializeValueMapper.java:
```diff
@@ -0,0 +1,185 @@
+        {
+            return value;
+        }
+        return ValueUtils.fromRelationshipProxy( defaultValueMapper.mapRelationship( value ) );
```
same here use the proxySpi instead

pontusmelke commented on community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledMaterializeValueMapper.java:
```diff
@@ -0,0 +1,185 @@
+        {
+            return value;
+        }
+        return ValueUtils.fromPath( defaultValueMapper.mapPath( value ) );
```
in this case can't we just alway return `value`?

pontusmelke commented on community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/NodeIdWrapperImpl.java:
```diff
@@ -20,8 +20,10 @@
+import org.neo4j.values.virtual.VirtualNodeValue;
 
-public final class NodeIdWrapperImpl implements NodeIdWrapper
+public final class NodeIdWrapperImpl extends VirtualNodeValue implements NodeIdWrapper
```
Can't we just use NodeReference directly and get rid of the [Node,Relationship]Wrapper altogether?

henriknyman replied:
Yes definitely. I didn't get to that yet though.

pontusmelke commented on community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/RelationshipIdWrapperImpl.java:
```diff
@@ -20,8 +20,10 @@
+import org.neo4j.values.virtual.VirtualRelationshipValue;
 
-public final class RelationshipIdWrapperImpl implements RelationshipIdWrapper
+public final class RelationshipIdWrapperImpl extends VirtualRelationshipValue implements RelationshipIdWrapper
```
Can't we just use NodeReference directly and get rid of the [Node,Relationship]Wrapper altogether?

henriknyman replied:
Yes definitely. I didn't get to that yet though.

pontusmelke commented on community/values/src/main/java/org/neo4j/values/storable/NoValue.java:
```diff
@@ -28,7 +28,7 @@
  * and avoid unnecessary object creation.
  */
-final class NoValue extends Value
+public final class NoValue extends Value
```
Why do we make this public? No one should rely on this class but only on `Values.NO_VALUE`.

henriknyman replied:
Hmm it looks like I used the eq-method in one place and an instanceof-check in another, but it was just being silly. I will replace with == and make this package private again.

pontusmelke reviewed (changes requested):
Nice work and nice test coverage. I had some comments, not sure all are valid but please look through them and see if you think they make sense.

henriknyman committed:
Fix review comments

- Do not use NoValue class directly
- Minor optimization of sequence values in toSet()

henriknyman committed:
Fix compilation error

Resolve conflict in test implementation after rebase

henriknyman committed:
Fix issue with generated comparators

If an integer, float or boolean comes as an AnyValue, but only has
the known representation of ReferenceType at compile time, we cannot
generate primitive comparators.
